src/transformers/models/dac/modeling_dac.py

```diff
@@ -0,0 +1,653 @@
+        return_dict: Optional[bool] = None,
+    ):
+        """
+        Encode given audio data and return quantized latent codes
+
+        Args:
+            input_values (`torch.Tensor of shape `(batch_size, 1, time_steps)`):
+                Input audio data to encode,
+            n_quantizers (int, *optional*):
+                Number of quantizers to use. If None, all quantizers are used. Default is None.
+            return_dict (`bool`, *optional*):
+                Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+        Returns:
+            loss (`torch.Tensor` of shape ((batch_size, 1, time_steps)):
+                Loss from the encoder model, comprising the weighted combination of the commitment and codebook losses.
+            quantized_representation (`torch.Tensor` of shape `(batch_size, dimension, time_steps)`):
+                Quantized continuous representation of input.
+            codebook_indices (`torch.Tensor` of shape `(batch_size, num_codebooks, time_steps)`):
+                Codebook indices for each codebook (quantized discrete representation of input),
+            projected_latents (`torch.Tensor` of shape `(batch_size, num_codebooks * dimension, time_steps)`):
+                Projected latents (continuous representation of input before quantization),
+        """
+        return_dict = return_dict if return_dict is not None else self.config.return_dict
+
+        quantized_representation = self.encoder(input_values)
+        quantized_representation, codebook_indices, projected_latents, commitment_loss, codebook_loss = self.quantizer(
+            quantized_representation, n_quantizers
+        )
+
+        loss = self.config.commitment_loss_weight * commitment_loss + self.config.codebook_loss_weight * codebook_loss
+
+        if not return_dict:
+            return (loss, quantized_representation, codebook_indices, projected_latents)
+
+        return DacEncoderOutput(loss, quantized_representation, codebook_indices, projected_latents)
+
+    def decode(
+        self,
+        quantized_representation: torch.Tensor,
+        return_dict: Optional[bool] = None,
+    ):
+        """Decode given latent codes and return audio data
+
+        Args:
+            quantized_representation (torch.Tensor of shape `(batch_size, dimension, time_steps)`):
+                Quantized continuous representation of input
+            return_dict (`bool`, *optional*):
+                Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+        Returns:
+            audio_values (`torch.Tensor` of shape `(batch_size, 1, input_length)`):
+                Decoded audio data.
+        """
+        return_dict = return_dict if return_dict is not None else self.config.return_dict
+
+        audio_values = self.decoder(quantized_representation)
+
+        if not return_dict:
+            return (audio_values,)
+
+        return DacDecoderOutput(audio_values)
```

Comment:
We should absolutely also be able to also decode from `audio_codes`, as it'll be the most common use-case !

I think it's done here in the original codebase: https://github.com/descriptinc/descript-audio-codec/blob/c7cfc5d2647e26471dc394f95846a0830e7bec34/dac/model/base.py#L236-L294

Reply:
This is also what is done in Encodec FYI ! https://github.com/sanchit-gandhi/parler-tts/blob/10016fb0300c0dc31a0fb70e26f3affee7b62f16/parler_tts/dac_wrapper/modeling_dac.py#L98-L134